lib/Services/cayenne.js

```diff
@@ -125,6 +125,90 @@ class Cayenne {
+    * Gets all attribute groups and attributes for a device
+    * @param {String} device_id The device id to get attributes for
+    */
+   getAttributeGroups(device_id) {
+     const url = `${this.url}/things/${device_id}/groups`;
+     return this.auth.send(this.service, 'GET', url, { query });
+   }
+
+   /**
+    * Create or Update an attribute group. Pass null to group_id to create a new group.
+    * @param {String} device_id The id of the parent device
+    * @param {String?} group_id Optional: Updates group, if null creates instead
+    * @param {Object} payload The group to create or update
+    * @param {UUID?} payload.id Optional: The group id, will be overriden by group_id parameter
+    * @param {String} payload.name The name of the group
+    * @param {Integer} payload.order The display order of the group
+    * @param {String?} payload.device_id Optional: The device_id this group is attached to. Overriden by device_id paramter
+    * @param {UUID} payload.user_id The user_id of the device owner
+    */
+   createOrUpdateGroup(device_id, group_id, payload) {
+     var url = null
+     if(group_id == null){
+       url = `${this.url}/things/${device_id}/groups`;
+     } else {
+       url = `${this.url}/things/${device_id}/groups/${group_id}`;
+     }
+     return this.auth.send(this.service, 'PUT', url, { payload });
+   }
+
+   /**
+    * Delete a group
+    * @param {String} device_id 
+    * @param {String} group_id 
+    */
+   deleteGroup(device_id, group_id) {
+     const url = `/things/${device_id}/groups/${group_id}`;
+     return this.auth.send(this.service, 'DELETE', url, { payload });
+   }
+
+   /**
+    * Create or Update an attribute. Pass null to attribute_id to create a new attribute.
+    * @param {String} device_id The id of the parent device
+    * @param {String} group_id The id of the parent group
+    * @param {String?} attribute_id Optional: Updates attribute, if null creates instead
+    * @param {Object} payload The attribute to create or update
+    * @param {UUID?} payload.id Optional: The attribute id, will be overriden by attribute_id parameter
+    * @param {String} payload.name The name of the attribute
+    * @param {Integer} payload.order The display order of the attribute
+    * @param {String?} payload.device_id Optional: The device_id this attribute is attached to. Overriden by device_id paramter
+    * @param {UUID} payload.user_id The user_id of the device owner
+    * @param {UUID?} payload.group_id The group_id of the parent group
+    * @param {String} payload.value The value, contents vary based on value_type
+    * @param {String} payload.value_type The type of content contained in value string
+    */
+   createOrUpdateAttribute(device_id, group_id, attribute_id, payload) {
+    var url = null
+    if(attribute_id == null){
+      url = `${this.url}/things/${device_id}/groups/${group_id}`;
+    } else {
+      url = `${this.url}/things/${device_id}/groups/${group_id}/${attribute_id}`;
```

Comment:
Is `attributes` missing from the path (also on L#204)?  `${this.url}/things/${device_id}/groups/${group_id}/attributes/${attribute_id}`

Reply:
yup :D